### PR TITLE
Update KNN

### PR DESCRIPTION
**Pull request by Saurabh7**
WIP

### CODE REVIEW

**Comment by Saurabh7** [comment]
a few results https://gist.github.com/Saurabh7/05369cb213290b82060859e7324fccac
Current verison of kdtree helps for lower dimensions as it should but a bit slower in comparison.

**Comment by vigsterkr** [comment]
@Saurabh7 quite big diff on the same data set (smaller dim) in case of SK... how many time do you run the tests? just to see what's the (mean, std) of runtimes....

**Comment by Saurabh7** [comment]
The difference is consistently in the same range when ran multiple times.
code : https://gist.github.com/Saurabh7/88d09aa530257a17a6f73b0b8c47a7b7

**Comment by Saurabh7** [comment]
Actually @vigsterkr there was something wrong. The leaf sizes were different for kdtree.
corrected version: https://gist.github.com/Saurabh7/b50a3043a98a1fa9b870ed5797486520

**Comment by karlnapf** [comment]
Can we also parallelise thisi eventually?

File: src/shogun/multiclass/KNN.cpp [hunk 1]
```diff
@@ -100,6 +102,12 @@ bool CKNN::train_machine(CFeatures* data)
 	m_min_label=min_class;
 	m_num_classes=max_class-min_class+1;
 
+	if (m_knn_solver == KNN_KDTREE)
+	{
+		m_kd_tree = new CKDTree(m_leaf_size);
```
Comment (karlnapf) [inline src/shogun/multiclass/KNN.cpp]: this is a well tested shogun class?
I remember either kdtree or cover tree was buggy
Also, are both implementation BSD compatible?

File: src/shogun/multiclass/KNN.cpp [hunk 2]
```diff
@@ -203,9 +213,10 @@ CMulticlassLabels* CKNN::apply_multiclass(CFeatures* data)
 #ifdef BENCHMARK_KNN
```
Comment (karlnapf) [inline src/shogun/multiclass/KNN.cpp]: can we pls remove these horrible things? :)

File: src/shogun/multiclass/KNN.cpp [hunk 3]
```diff
@@ -277,6 +288,25 @@ CMulticlassLabels* CKNN::apply_multiclass(CFeatures* data)
...
 		SG_PRINT(">>>> JL applied in %9.4f\n", tstart.cur_time_diff(false))
 #endif
+		break;
+	}
+	case KNN_KDTREE:
+	{
+		m_kd_tree->query_knn(dynamic_cast<CDenseFeatures<float64_t>*>(data), m_k);
+		SGMatrix<index_t> NN = m_kd_tree->get_knn_indices();
+		for (int32_t i=0; i<num_lab && (!CSignal::cancel_computations()); i++)
```
Comment (karlnapf) [inline src/shogun/multiclass/KNN.cpp]: This thing is actually quite good to have. We should put this into other algos.
Wanna send a snother patch for that?
I think it doesnt work for openmp though

File: src/shogun/multiclass/KNN.cpp [hunk 4]
```diff
@@ -31,7 +31,7 @@ CKNN::CKNN()
 	init();
 }
 
-CKNN::CKNN(int32_t k, CDistance* d, CLabels* trainlab)
+CKNN::CKNN(int32_t k, CDistance* d, CLabels* trainlab, KNN_SOLVER_TYPE knn_solver)
```
Comment (karlnapf) [inline src/shogun/multiclass/KNN.cpp]: We should aim for mimicing sklearns setup and default parameters, as I think it is really well tuned

**Comment by karlnapf** [comment]
Why are we slower than sklearn? Any hunches? Looked at their code?

**Comment by Saurabh7** [comment]
Brute force should be fixed with some precomputations I think.

File: src/shogun/multiclass/KNN.cpp (hunk 1 above)
Comment (Saurabh7) [inline src/shogun/multiclass/KNN.cpp]: There are tests, more might be needed I will check

File: src/shogun/machine/DistanceMachine.cpp [hunk 5]
```diff
@@ -56,59 +56,10 @@ void CDistanceMachine::distances_lhs(float64_t* result,int32_t idx_a1,int32_t id
...
-    if (num_threads < 2)
-    {
-        D_THREAD_PARAM param;
-        param.d=distance;
-        param.r=result;
-        param.idx_r_start=idx_a1;
-        param.idx_start=idx_a1;
-        param.idx_stop=idx_a2+1;
-        param.idx_comp=idx_b;
+#pragma omp parallel for
```
Comment (karlnapf) [inline src/shogun/machine/DistanceMachine.cpp]: this is quite a big change.
Rationale?

**Comment by karlnapf** [comment]
Mmmh this is getting too big, we need to chop it into pieces. Can you send smaller PRs with some explanation what you are doing?

**Comment by Saurabh7** [comment]
Yep, I have added falconn lib external code here too (https://github.com/FALCONN-LIB/FALCONN), for LSH based approximate KNN. I will do that in separate PR then. 
Need to fix some tests and find a good way to set their parameters.  To be able to control tradeoff between accuracy and speed. 
for example for a particular set of parameters (not the best i think) I get on this data

```
Dataset    Shape            SG time    SK time    ACC sg    ACC sk
---------  -------------  ---------  ---------  --------  --------
toy        [10000, 1000]    1.19713    3.50962     0.943     0.987

```

**Comment by karlnapf** [comment]
Cool stuff. 
But the KD tree is also necessary, and a general clean up.
Please do the patches one after another as small as possible so that we can review easier

**Comment by Saurabh7** [comment]
Yes, KD tree is already in.

**Comment by Saurabh7** [comment]
added tests too, could be good to go ..

File: src/shogun/multiclass/KNN.cpp [hunk 6]
```diff
@@ -100,6 +102,14 @@ bool CKNN::train_machine(CFeatures* data)
 	m_min_label=min_class;
 	m_num_classes=max_class-min_class+1;
 
+	if (m_knn_solver == KNN_KDTREE)
+	{
+		CFeatures* lhs = distance->get_lhs();
+		m_kd_tree = new CKDTree(m_leaf_size);
+		m_kd_tree->build_tree(dynamic_cast<CDenseFeatures<float64_t>*>(lhs));
```
Comment (vigsterkr) [inline src/shogun/multiclass/KNN.cpp]: our current kd tree only works with `Densefeature<float64_t>`?

Comment (Saurabh7) [inline src/shogun/multiclass/KNN.cpp]: Yup looks like it.

File: src/shogun/machine/DistanceMachine.cpp (hunk 5 above)
Comment (Saurabh7) [inline src/shogun/machine/DistanceMachine.cpp]: I am removing this too, not absolutely necessary, can discuss later.

**Comment by vigsterkr** [comment]
@Saurabh7 great! can you please squash this so i can merge it?
and in the next PR send in the falconn based update?

**Comment by Saurabh7** [comment]
Squased! waiting for travis ..

**Comment by vigsterkr** [comment]
@Saurabh7 ok cool as soon as it's green we can merge.

**Pull request merged**